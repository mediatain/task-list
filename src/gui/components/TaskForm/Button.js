import styled, { css } from 'styled-components';
import { taskItemColor, textColorDark } from '../../colors';
import { elevation2, elevation8 } from '../../elevations';

export default styled.button.attrs({
  type: 'button',
})`
  background: ${taskItemColor};
  color: ${textColorDark};
  padding: 0.7rem 1.2rem;
  font-size: 1rem;
  font-weight: 500;
  border-radius: 5px;
  text-transform: uppercase;
  border: none;
  elevation: above;
  cursor: pointer;
  white-space: nowrap;
  outline: none;
  box-shadow: ${elevation2};
  transition-property: box-shadow, opacity;
  transition: 150ms ease-in-out;
  
  &:hover {
    box-shadow: ${elevation8};
  }
  
  ${({ invisible }) => invisible && css`
    opacity: 0;
    overflow: hidden;
  `}
`;
