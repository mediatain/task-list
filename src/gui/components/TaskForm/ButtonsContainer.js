import styled from 'styled-components';

export default styled.div`
  display: flex;
  margin-left: .7rem;

  @media (max-width: 38rem) {
    flex-direction: column;
    justify-content: stretch;
  }
`;
