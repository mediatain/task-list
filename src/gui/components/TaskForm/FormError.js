import React from 'react';
import styled from 'styled-components';
import { warningColor } from '../../colors';

const StyledFormError = styled.div`
  text-align: center;
  margin-top: .5rem;
  padding-top: 1rem;
  color: ${warningColor};
  font-size: 1.5rem;
`;

export default function FormError({ message }) {
  return (
    <StyledFormError>
      <span role="img" aria-label="Warning">⚠️</span>
      <span> {message}</span>
    </StyledFormError>
  );
}
