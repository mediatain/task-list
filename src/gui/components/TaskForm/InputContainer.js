import styled, { css } from 'styled-components';

export default styled.div`
  position: absolute;
  display: flex;
  width: 0;
  overflow: hidden;
  opacity: 0;
  transition-property: width, opacity;
  transition: 150ms ease-in-out;

  ${({ visible }) => visible && css`
    width: 100%;
    opacity: 1;
    overflow: visible;
  `}
`;
