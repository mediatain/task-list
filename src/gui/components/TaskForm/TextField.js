import styled from 'styled-components';
import { disabledColor, taskItemColor, textColorDark } from '../../colors';

export default styled.input`
  flex: 1;
  outline: none;
  color: ${textColorDark};
  padding: 0.5rem;
  font-size: 1.5rem;
  border-top: none;
  border-left: none;
  border-right: none;
  background: none;
  border-bottom: .2rem solid ${disabledColor};
  width: 0;

  &:focus {
    border-bottom: .2rem solid ${taskItemColor};
  }
`;
