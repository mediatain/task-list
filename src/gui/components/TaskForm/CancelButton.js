import styled from 'styled-components';
import Button from './Button';
import { cancelColor } from '../../colors';

export default styled(Button).attrs({
  type: 'reset',
})`
  background: ${cancelColor};

  @media (max-width: 38rem) {
    padding: .25rem 1rem;
    font-size: 1rem;
    flex-grow: 1;
  }
`;
