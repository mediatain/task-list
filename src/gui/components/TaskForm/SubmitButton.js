import styled from 'styled-components';
import Button from './Button';
import { confirmColor } from '../../colors';

export default styled(Button).attrs({
  type: 'submit',
})`
  background: ${confirmColor};
  margin-left: .5rem;

  @media (max-width: 38rem) {
    padding: .25rem 1rem;
    font-size: 1rem;
    flex-grow: 1;
    margin-left: 0;
    margin-top: .5rem;
  }
`;
