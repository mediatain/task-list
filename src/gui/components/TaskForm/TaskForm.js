import React, { useRef, useState } from 'react';
import FormError from './FormError';
import CancelButton from './CancelButton';
import SubmitButton from './SubmitButton';
import ButtonsContainer from './ButtonsContainer';
import Button from './Button';
import styled from 'styled-components';
import TextField from './TextField';
import InputContainer from './InputContainer';
import { cancelButtonLabel, newTaskButtonLabel, saveButtonLabel } from '../../strings';

const StyledTaskForm = styled.form`
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  max-width: 40rem;
  margin: auto;
  padding: 0;
`;

export default function TaskForm({ onAddTask }) {
  const textInputRef = useRef();
  const [value, setValue] = useState('');
  const [open, setOpen] = useState(false);
  const [error, setError] = useState();

  function resetForm() {
    setError();
    setOpen(false);
    setValue('');
    if (textInputRef.current) {
      textInputRef.current.blur();
    }
  }

  function onOpen() {
    setOpen(true);
    if (textInputRef.current) {
      textInputRef.current.focus();
    }
  }

  function onSubmit(e) {
    e.preventDefault();
    try {
      onAddTask(value);
      resetForm();
    } catch (e) {
      setError(e.message);
    }
  }

  function onChangeValue(e) {
    setValue(e.target.value);
    setError(undefined);
  }

  return (
    <StyledTaskForm onSubmit={onSubmit}>
      <Button onClick={onOpen} invisible={open}>{newTaskButtonLabel}</Button>
      <InputContainer visible={open}>
        <TextField
          ref={textInputRef}
          value={value}
          onChange={onChangeValue}
        />
        <ButtonsContainer>
          <CancelButton onClick={resetForm}>{cancelButtonLabel}</CancelButton>
          <SubmitButton>{saveButtonLabel}</SubmitButton>
        </ButtonsContainer>
      </InputContainer>
      {error && <FormError message={error} />}
    </StyledTaskForm>
  );
}
