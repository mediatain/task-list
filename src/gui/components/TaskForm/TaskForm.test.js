import React from 'react';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import TaskForm from './TaskForm';
import { shallow } from 'enzyme';
import Button from './Button';
import FormRow from './InputContainer';
import FormInput from './TextField';
import FormError from './FormError';
import { cancelButtonLabel } from '../../strings';

function submit(taskForm) {
  taskForm.simulate('submit', { preventDefault: jest.fn() });
}

function cancel(taskForm) {
  const cancelButton = taskForm.find({ children: cancelButtonLabel });
  cancelButton.simulate('click');
}

function changeText(taskForm, inputValue) {
  const formInput = taskForm.find(FormInput);
  const changeEvent = { target: { value: inputValue } };
  formInput.simulate('change', changeEvent);
}

describe('TaskForm', () => {
  it('should match the snapshot', () => {
    const { container } = render(<TaskForm onAddTask={jest.fn()} />);

    expect(container).toMatchSnapshot();
  });

  it('should not show the input by default', () => {
    const taskForm = shallow(<TaskForm onAddTask={jest.fn()} />);

    const formRow = taskForm.find(FormRow);
    expect(formRow.prop('visible')).toBe(false);
  });

  it('should show the input when New Task Button is clicked', () => {
    const taskForm = shallow(<TaskForm onAddTask={jest.fn()} />);

    taskForm.find(Button).simulate('click');

    const formRow = taskForm.find(FormRow);
    expect(formRow.prop('visible')).toBe(true);
  });

  describe('when submitting', () => {
    it('should call the passed onAddTask function with entered text', () => {
      const onAddTask = jest.fn();
      const inputValue = 'My task';
      const taskForm = shallow(<TaskForm onAddTask={onAddTask} />);
      changeText(taskForm, inputValue);

      submit(taskForm);

      expect(onAddTask).toBeCalledWith(inputValue);
    });

    it('should clear the entered text', () => {
      const taskForm = shallow(<TaskForm onAddTask={jest.fn()} />);
      changeText(taskForm, 'My task');

      submit(taskForm);

      const formInput = taskForm.find(FormInput);
      expect(formInput.prop('value')).toBe('');
    });

    it('should close the input', () => {
      const taskForm = shallow(<TaskForm onAddTask={jest.fn()} />);
      changeText(taskForm, 'My task');

      submit(taskForm);

      const formRow = taskForm.find(FormRow);
      expect(formRow.prop('visible')).toBe(false);
    });

    describe('when onAddTask throws', () => {
      it('should show an error', () => {
        const onAddTask = jest.fn(() => {throw new Error('An error occurred.');});
        const taskForm = shallow(<TaskForm onAddTask={onAddTask} />);

        submit(taskForm);

        expect(taskForm.exists(FormError)).toBe(true);
      });

      it('should display the error message', () => {
        const errorMessage = 'An error occurred.';
        const onAddTask = jest.fn(() => {throw new Error(errorMessage);});
        const taskForm = shallow(<TaskForm onAddTask={onAddTask} />);

        submit(taskForm);

        const formError = taskForm.find(FormError).dive();
        expect(formError.contains(errorMessage)).toBe(true);
      });
    });
  });

  describe('when canceling', () => {
    it('should clear the entered text', () => {
      const taskForm = shallow(<TaskForm onAddTask={jest.fn()} />);
      changeText(taskForm, 'My task');

      cancel(taskForm);

      const formInput = taskForm.find(FormInput);
      expect(formInput.prop('value')).toBe('');
    });

    it('should close the input', () => {
      const taskForm = shallow(<TaskForm onAddTask={jest.fn()} />);
      changeText(taskForm, 'My task');

      cancel(taskForm);

      const formRow = taskForm.find(FormRow);
      expect(formRow.prop('visible')).toBe(false);
    });

    it('should hide an error message when there was one', () => {
      const onAddTask = jest.fn(() => {throw new Error('An error occurred.');});
      const taskForm = shallow(<TaskForm onAddTask={onAddTask} />);
      submit(taskForm);
      expect(taskForm.exists(FormError)).toBe(true);

      cancel(taskForm);

      expect(taskForm.exists(FormError)).toBe(false);
    });
  });

  describe('when typing', () => {
    it('should hide an error message when there was one', () => {
      const onAddTask = jest.fn(() => {throw new Error('An error occurred.');});
      const taskForm = shallow(<TaskForm onAddTask={onAddTask} />);
      submit(taskForm);
      expect(taskForm.exists(FormError)).toBe(true);

      changeText(taskForm);

      expect(taskForm.exists(FormError)).toBe(false);
    });
  });
});

