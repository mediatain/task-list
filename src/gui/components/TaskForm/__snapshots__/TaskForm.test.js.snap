// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm should match the snapshot 1`] = `
.c1 {
  background: #fdd835;
  color: #000000;
  padding: 0.7rem 1.2rem;
  font-size: 1rem;
  font-weight: 500;
  border-radius: 5px;
  text-transform: uppercase;
  border: none;
  elevation: above;
  cursor: pointer;
  white-space: nowrap;
  outline: none;
  box-shadow: 0 2px 2px 0 rgba(0,0,0,0.14),0 3px 1px -2px rgba(0,0,0,0.12),0 1px 5px 0 rgba(0,0,0,0.20);
  -webkit-transition-property: box-shadow,opacity;
  transition-property: box-shadow,opacity;
  -webkit-transition: 150ms ease-in-out;
  transition: 150ms ease-in-out;
}

.c1:hover {
  box-shadow: 0 8px 10px 1px rgba(0,0,0,0.14),0 3px 14px 2px rgba(0,0,0,0.12),0 5px 5px -3px rgba(0,0,0,0.20);
}

.c5 {
  background: #fdd835;
  color: #000000;
  padding: 0.7rem 1.2rem;
  font-size: 1rem;
  font-weight: 500;
  border-radius: 5px;
  text-transform: uppercase;
  border: none;
  elevation: above;
  cursor: pointer;
  white-space: nowrap;
  outline: none;
  box-shadow: 0 2px 2px 0 rgba(0,0,0,0.14),0 3px 1px -2px rgba(0,0,0,0.12),0 1px 5px 0 rgba(0,0,0,0.20);
  -webkit-transition-property: box-shadow,opacity;
  transition-property: box-shadow,opacity;
  -webkit-transition: 150ms ease-in-out;
  transition: 150ms ease-in-out;
  background: #EC5555;
}

.c5:hover {
  box-shadow: 0 8px 10px 1px rgba(0,0,0,0.14),0 3px 14px 2px rgba(0,0,0,0.12),0 5px 5px -3px rgba(0,0,0,0.20);
}

.c6 {
  background: #fdd835;
  color: #000000;
  padding: 0.7rem 1.2rem;
  font-size: 1rem;
  font-weight: 500;
  border-radius: 5px;
  text-transform: uppercase;
  border: none;
  elevation: above;
  cursor: pointer;
  white-space: nowrap;
  outline: none;
  box-shadow: 0 2px 2px 0 rgba(0,0,0,0.14),0 3px 1px -2px rgba(0,0,0,0.12),0 1px 5px 0 rgba(0,0,0,0.20);
  -webkit-transition-property: box-shadow,opacity;
  transition-property: box-shadow,opacity;
  -webkit-transition: 150ms ease-in-out;
  transition: 150ms ease-in-out;
  background: #2b9949;
  margin-left: .5rem;
}

.c6:hover {
  box-shadow: 0 8px 10px 1px rgba(0,0,0,0.14),0 3px 14px 2px rgba(0,0,0,0.12),0 5px 5px -3px rgba(0,0,0,0.20);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: .7rem;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  outline: none;
  color: #000000;
  padding: 0.5rem;
  font-size: 1.5rem;
  border-top: none;
  border-left: none;
  border-right: none;
  background: none;
  border-bottom: .2rem solid #a4a5a4;
  width: 0;
}

.c3:focus {
  border-bottom: .2rem solid #fdd835;
}

.c2 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 0;
  overflow: hidden;
  opacity: 0;
  -webkit-transition-property: width,opacity;
  transition-property: width,opacity;
  -webkit-transition: 150ms ease-in-out;
  transition: 150ms ease-in-out;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  max-width: 40rem;
  margin: auto;
  padding: 0;
}

@media (max-width:38rem) {
  .c5 {
    padding: .25rem 1rem;
    font-size: 1rem;
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
  }
}

@media (max-width:38rem) {
  .c6 {
    padding: .25rem 1rem;
    font-size: 1rem;
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    margin-left: 0;
    margin-top: .5rem;
  }
}

@media (max-width:38rem) {
  .c4 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: stretch;
    -webkit-justify-content: stretch;
    -ms-flex-pack: stretch;
    justify-content: stretch;
  }
}

<div>
  <form
    class="c0"
  >
    <button
      class="c1"
      type="button"
    >
      New task
    </button>
    <div
      class="c2"
    >
      <input
        class="c3"
        value=""
      />
      <div
        class="c4"
      >
        <button
          class="c5"
          type="reset"
        >
          Cancel
        </button>
        <button
          class="c6"
          type="submit"
        >
          Save
        </button>
      </div>
    </div>
  </form>
</div>
`;
